core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/ldap/LDAPConnectionContext.java

```diff
@@ -429,4 +461,191 @@ private LdapContext getContextForEnvironmentVariables(Hashtable<?, ?> environmen
+            correlationLogDTO.setArgs(CORRELATION_LOG_INITIALIZATION_ARGS);
+            logDetails(correlationLogDTO);
+            return (LdapContext) proxy;
+        } else {
+            return new InitialLdapContext(environment, connectionControls);
+        }
+    }
+
+    /**
+     * Proxy Class that is used to calculate and log the time taken for queries
+     */
+    private class LdapContextInvocationHandler implements InvocationHandler {
+
+        private Object previousContext;
+
+        public LdapContextInvocationHandler(Object previousContext) {
+
+            this.previousContext = previousContext;
+        }
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+
+            long start = System.currentTimeMillis();
+            Object result = method.invoke(this.previousContext, args);
+            long delta = System.currentTimeMillis() - start;
+            String methodName = method.getName();
+            int argsLength = 0;
+
+            if (args != null) {
+                argsLength = args.length;
+            }
+
+            if (!StringUtils.equalsIgnoreCase("close", methodName)) {
+                CorrelationLogDTO correlationLogDTO = new CorrelationLogDTO();
+                correlationLogDTO.setStartTime(start);
+                correlationLogDTO.setDelta(delta);
+                correlationLogDTO.setEnvironment(((DirContext) this.previousContext).getEnvironment());
+                correlationLogDTO.setMethodName(methodName);
+                correlationLogDTO.setArgsLength(argsLength);
+                correlationLogDTO.setArgs(stringify(args));
+                logDetails(correlationLogDTO);
+            }
+            return result;
+        }
+
+        /**
+         * Creates a argument string by appending the values in the array
+         *
+         * @param arr Arguments
+         * @return Argument string
+         */
+        private String stringify(Object[] arr) {
+
+            StringBuilder sb = new StringBuilder();
+            if (arr == null) {
+                sb.append("null");
+            } else {
+                sb.append(" ");
+                for (int i = 0; i < arr.length; i++) {
+                    Object o = arr[i];
+                    sb.append(o.toString());
+                    if (i < arr.length - 1) {
+                        sb.append(",");
+                    }
+                }
+            }
+            return sb.toString();
+        }
+    }
+
+    /**
+     * Logs the details from the LDAP query
+     *
+     * @param correlationLogDTO Contains all details that should be to logged
+     */
+    private void logDetails(CorrelationLogDTO correlationLogDTO) {
+
+        String providerUrl = "No provider url found";
+        String principal = "No principal found";
+
+        if (correlationLogDTO.getEnvironment().containsKey("java.naming.provider.url")) {
```

Comment:
shall we define constants for "java.naming.provider.url" and  "java.naming.security.principal" ?